### PR TITLE
Workspace server init

### PR DESCRIPTION
**Pull request by polohan**

### CODE REVIEW

File: lib/workspaceSocket.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 const socketServer = require('./socket-server');
+const request = require('request')
```
Comment (tdy) [inline lib/workspaceSocket.js:2]:
```suggestion
const request = require('request');
```

File: lib/workspaceSocket.js [hunk 2]
```diff
@@ -9,22 +10,59 @@ module.exports.init = function (callback) {
   callback(null);
 };
 
+async function controlContainer(callback, action) {
+    let promise = new Promise(resolve => {
+        request.post('http://localhost:8081', {
+            json: {
+                workspace_id: 0,
+                action: action
```
Comment (tdy) [inline lib/workspaceSocket.js:18]:
```suggestion
                action: action,
```

File: lib/workspaceSocket.js [hunk 3]
```diff
@@ -9,22 +10,59 @@ module.exports.init = function (callback) {
...
+                console.log(error);
+                return;
+            }
+            console.log(`statusCode: ${response.statusCode}`);
+            console.log(body);
+            if (response.statusCode == 200) {
+                resolve(true);
+            } else {
+                resolve(false);
+            };
```
Comment (tdy) [inline lib/workspaceSocket.js:31]:
```suggestion
            }
```

File: lib/workspaceSocket.js [hunk 4]
```diff
@@ -9,22 +10,59 @@ module.exports.init = function (callback) {
...
+    let res = await promise;
+    if (res) {
+        if (action == "init") {
+            console.log("Container started");
+            callback.emit('host-ready', "");
+        } else if (action == "stop") {
+            console.log("Container stopped");
+        }
+    } else {
+        console.log("Failed to execuate " + action);
```
Comment (tdy) [inline lib/workspaceSocket.js:43]:
```suggestion
        if (action == 'init') {
            console.log('Container started');
            callback.emit('host-ready', '');
        } else if (action == 'stop') {
            console.log('Container stopped');
        }
    } else {
        console.log('Failed to execuate ' + action);
```

File: lib/workspaceSocket.js [hunk 5]
```diff
@@ -9,22 +10,59 @@ module.exports.init = function (callback) {
...
+    };
+};
```
Comment (tdy) [inline lib/workspaceSocket.js:45]:
```suggestion
    }
}
```

File: lib/workspaceSocket.js [hunk 6]
```diff
@@ -9,22 +10,59 @@ module.exports.init = function (callback) {
...
 module.exports.connection = function (socket) {
     var workspace_id = socket.handshake.query['id'];
 
     socket.join(workspace_id, () => {
         let rooms = Object.keys(socket.rooms);
         console.log(rooms);
         console.log('Connection to workspace ' + workspace_id + ' established.');
-        module.exports._namespace.to(workspace_id).emit('update', workspace_id);
+        console.log('Initlizing remote workspace host');
+        controlContainer(module.exports._namespace.to(workspace_id), "init");
```
Comment (tdy) [inline lib/workspaceSocket.js:55]:
```suggestion
        controlContainer(module.exports._namespace.to(workspace_id), 'init');
```

File: lib/workspaceSocket.js [hunk 7]
```diff
@@ -9,22 +10,59 @@ module.exports.init = function (callback) {
...
     });
 
     socket.on('heartbeat', function () {
         console.log('Ping from workspace ' + workspace_id + ' received.');
+        controlContainer(module.exports._namespace.to(workspace_id), "sync");
```
Comment (tdy) [inline lib/workspaceSocket.js:60]:
```suggestion
        controlContainer(module.exports._namespace.to(workspace_id), 'sync');
```

File: lib/workspaceSocket.js [hunk 8]
```diff
@@ -9,22 +10,59 @@ module.exports.init = function (callback) {
...
         module.exports._namespace.to(workspace_id).emit('update', 'I got your msg!');
     });
 
     socket.on('disconnect', () => {
         console.log('user disconnected');
+        controlContainer(module.exports._namespace.to(workspace_id), "stop");
```
Comment (tdy) [inline lib/workspaceSocket.js:66]:
```suggestion
        controlContainer(module.exports._namespace.to(workspace_id), 'stop');
```

**Comment by nicknytko** [comment]
Notes from earlier to get this working:

- set bucket name in `lib/config.js`
- (if running in docker) change `localhost` to `host.docker.internal` in `server.js` and `lib/workspaceSocket.js`
- make sure image is pulled
- run `node interface.js` in the workspace subdirectory

**Comment by tdy** [comment]
> run `node interface.js` in the workspace subdirectory

I think it needs to be run from the root dir (where `aws-config.json` is located)?

**Comment by nicknytko** [comment]
> > run `node interface.js` in the workspace subdirectory
> 
> I think it needs to be run from the root dir (where `aws-config.json` is located)?

Yep, updated. Thanks.

**Comment by tdy** [comment]
Pasting @nicknytko's trick from slack -- stream s3 contents to stdout:

`aws s3 cp s3://<bucket>/workspace-0/<filename> -`

File: pages/workspace/workspace.ejs [hunk 9]
```diff
@@ -1,30 +1,72 @@
...
+      <div id="loading_page"><%- include('../partials/loading'); %></div>
+      <iframe
+        id="workspace_container"
+        style="margin: 0; border: none; width: 100%; height: calc(100% - 57px);"
+      ></iframe>
+    </div>
+    <button id="button" style="position: absolute; left: 250px; bottom: 0;">
+      Ping the Server
+    </button>
     <script src="/javascripts/socket.io.js"></script>
```
Comment (nicknytko) [inline pages/workspace/workspace.ejs:24]:
```suggestion
    <script src="/node_modules/socket.io-client/dist/socket.io.js"></script>
```

Local copy of `socket.io` was removed in https://github.com/PrairieLearn/PrairieLearn/commit/76f580ce55000ea0e13531735e00bc5d359ab4e3 in favor of the copy in node_modules.

File: lib/workspaceSocket.js [hunk 10]
```diff
@@ -9,22 +13,63 @@ module.exports.init = function (callback) {
...
 module.exports.connection = function (socket) {
-    var workspace_id = socket.handshake.query['id'];
+    const workspace_id = socket.handshake.query['workspace_id'];
 
     socket.join(workspace_id, () => {
         let rooms = Object.keys(socket.rooms);
         console.log(rooms);
         console.log('Connection to workspace ' + workspace_id + ' established.');
-        module.exports._namespace.to(workspace_id).emit('update', workspace_id);
+        console.log('Initlizing remote workspace host');
```
Comment (tdy) [inline lib/workspaceSocket.js:61]:
```suggestion
        console.log('Initializing remote workspace host');
```

File: workspace_host/interface.js [hunk 11]
```diff
@@ -244,7 +244,7 @@ async function _syncPushContainer(workspace_id, callback) {
         callback(null, workspace_id);
         return;
     }
-    var jobs_params = _recursiveUploadJobManager(workspaceName, workspaceName);
+    var jobs_params = _recursiveUploadJobManager(`/jobs/${workspaceName}`, workspaceName);
```
Comment (tdy) [inline workspace_host/interface.js:247]: @mwest1066 I just realized, we should only prepend `/jobs` if we have $HOST_JOBS_DIR, right?

Comment (tdy) [reply workspace_host/interface.js]: I modified it to this: https://github.com/PrairieLearn/PrairieLearn/pull/2723/commits/5f16e6a7ffad7222ef34bf38a5915f16d5950cc6

Comment (mwest1066) [reply workspace_host/interface.js]: Yep, that looks correct to me.

**Pull request merged**